diff --git a/build/modern/createPersister.js b/build/modern/createPersister.js
index d55afe42e5a0f890a0b54ab3b728fa68055794f4..51286d20cac69110c50b9e255835d53cbdcdce96 100644
--- a/build/modern/createPersister.js
+++ b/build/modern/createPersister.js
@@ -103,7 +103,8 @@ function experimental_createQueryPersister({
             dataUpdatedAt: persistedQuery.state.dataUpdatedAt,
             errorUpdatedAt: persistedQuery.state.errorUpdatedAt
           });
-          if (refetchOnRestore === "always" || refetchOnRestore === true && query.isStale()) {
+          const staleTime = typeof query.options.staleTime === "function" ? query.options.staleTime(query) : query.options.staleTime;
+          if (refetchOnRestore === "always" || refetchOnRestore === true && (query.isStale() || query.isStaleByTime(staleTime))) {
             query.fetch();
           }
         }
diff --git a/src/createPersister.ts b/src/createPersister.ts
index 16fa8c43104a6adaf545721921dddec99ed8fe52..a58cd2122b798503d1c7e7e53dbe573cd3ca4e4e 100644
--- a/src/createPersister.ts
+++ b/src/createPersister.ts
@@ -218,9 +218,14 @@ export function experimental_createQueryPersister<TStorageValue = string>({
             errorUpdatedAt: persistedQuery.state.errorUpdatedAt,
           })

+          const staleTime =
+            typeof query.options.staleTime === 'function'
+              ? query.options.staleTime(query)
+              : query.options.staleTime
           if (
             refetchOnRestore === 'always' ||
-            (refetchOnRestore === true && query.isStale())
+            (refetchOnRestore === true &&
+              (query.isStale() || query.isStaleByTime(staleTime)))
           ) {
             query.fetch()
           }
